Fix bug where updated webhook options failed to save (fixes #12336)

robwhitaker opened the pull request:
Fixes #12336

### Changes

This bug was caused by Mongoose not being able to tell when `Mixed` schema
objects (such as the `webhook.options` field) have been modified
(https://mongoosejs.com/docs/schematypes.html#mixed). So, although the webhook
was being updated properly, Mongoose was not actually committing it to the database.
Telling Mongoose that the webhook's `options` field has changed via `markModified`
fixes the issue.

Additionally, the relevant API test case was only checking whether or
not the webhook returned from the PUT endpoint matched the expected
update. Since the endpoint was returning the updated webhook without
querying the database again, this test case would pass. It has been
updated to check both the returned webhook as well as the version of the
webhook that is saved to the database against the expected. In other
words:

`assert returned === saved === expected`

----
UUID: cab16cfa-e951-4dc3-a468-1abadc1dd109

robwhitaker committed:
Fix bug where updated webhook options failed to save

This bug was caused by Mongoose not creating getters/setters for array
elements (https://mongoosejs.com/docs/faq.html#array-changes-not-saved).
So, although the webhook was being updated properly, Mongoose was not
actually committing it to the database. Telling Mongoose that the array
of webhooks has changed via `markModified` fixes the issue.

Additionally, the relevant API test case was only checking whether or
not the webhook returned from the PUT endpoint matched the expected
update. Since the endpoint was returning the updated webhook without
querying the database again, this test case would pass. It has been
updated to check both the returned webhook as well as the version of the
webhook that is saved to the database against the expected. In other
words:

`assert returned === saved === expected`

Fixes #12336

paglias commented on website/server/controllers/api-v3/webhook.js:
```diff
@@ -223,6 +223,10 @@ api.updateWebhook = {
 
+    // Tell Mongoose that the webhooks array has been modified
+    // so it actually commits the options changes to the database
+    user.markModified('webhooks');
```
The fix looks good but this way it's going to send back to the database the entire array even if only one item was modified, it works also if you do `webhook.markModified('options');` this way it's telling mongoose that the Mixed type `options` object is modified. The issue happens both with arrays but also mixed types objects as mongoose doesn't know what's been modified, see https://mongoosejs.com/docs/schematypes.html#mixed

robwhitaker replied:
Ah, okay. I've updated it to use `webhook.markModified('options')` instead. Also, I've updated the PR description to reflect the correction.

paglias replied:
thanks @robwhitaker ! The pr looks good and will be merged at the end of the week

paglias reviewed (changes requested):
@robwhitaker thanks for the PR! it looks good apart from the comment I left

robwhitaker committed:
Call markModified on webhook.options instead of user.webhooks

This tells Mongoose that only the modified webhook's options changed
instead of telling it that the entire user.webhooks array changed,
saving a costly DB update.

paglias commented:
thanks @robwhitaker ! I've noted this in your profile towards your 3rd tier and added the Blacksmith title
